devmand/gateway/src/devmand/channels/cli/TimeoutTrackingCli.cpp

```diff
@@ -0,0 +1,87 @@
+// Copyright (c) 2019-present, Facebook, Inc.
+// All rights reserved.
+//
+// This source code is licensed under the BSD-style license found in the
+// LICENSE file in the root directory of this source tree. An additional grant
+// of patent rights can be found in the PATENTS file in the same directory.
+
+#include <devmand/channels/cli/Command.h>
+#include <devmand/channels/cli/TimeoutTrackingCli.h>
+#include <magma_logging.h>
+
+namespace devmand::channels::cli {
+
+using namespace std;
+using namespace folly;
+using devmand::channels::cli::Command;
+
+shared_ptr<TimeoutTrackingCli> TimeoutTrackingCli::make(
+    string id,
+    shared_ptr<Cli> cli,
+    shared_ptr<folly::ThreadWheelTimekeeper> timekeeper,
+    shared_ptr<folly::Executor> executor,
+    std::chrono::milliseconds timeoutInterval) {
+  return shared_ptr<TimeoutTrackingCli>(
+      new TimeoutTrackingCli(id, cli, timekeeper, executor, timeoutInterval));
+}
+
+TimeoutTrackingCli::TimeoutTrackingCli(
+    string _id,
+    shared_ptr<Cli> _cli,
+    shared_ptr<folly::ThreadWheelTimekeeper> _timekeeper,
+    shared_ptr<folly::Executor> _executor,
+    std::chrono::milliseconds _timeoutInterval)
+    : id(_id),
+      cli(_cli),
+      timekeeper(_timekeeper),
+      executor(_executor),
+      timeoutInterval(_timeoutInterval) {
+  shutdown = false;
+}
+
+TimeoutTrackingCli::~TimeoutTrackingCli() {
+  MLOG(MDEBUG) << "[" << id << "] "
+               << "~TTCli";
+  shutdown = true;
+  executor = nullptr;
+  timekeeper = nullptr;
+  cli = nullptr;
+  MLOG(MDEBUG) << "[" << id << "] "
+               << "~TTCli done";
+}
+
+Future<string> TimeoutTrackingCli::executeAndRead(const Command& cmd) {
+  return executeSomething(cmd, "TTCli.executeAndRead", [this, cmd]() {
+    return cli->executeAndRead(cmd);
+  });
+}
+
+Future<string> TimeoutTrackingCli::execute(const Command& cmd) {
+  return executeSomething(
+      cmd, "TTCli.execute", [this, cmd]() { return cli->execute(cmd); });
+}
+
+Future<string> TimeoutTrackingCli::executeSomething(
+    const Command& cmd,
+    const string&& loggingPrefix,
+    const function<Future<string>()>& innerFunc) {
+  string cmdString = cmd.toString();
```

Comment:
cmdString is subsequenlty only used in "<<". Command supports direct call with "<<" so this varcould be omitted.